### ISSUE TITLE
Support non-root user and tools installations better

### ISSUE THREAD

**Issue by shalomb** [alexellis/arkade#363]
## Expected Behaviour

It's not always possible to install tools like `arkade` into the system directories via `sudo`, etc - secure environments.
Most "sensible" tools (`go get`, `pip`, `brew`, `cargo`, etc) detect if they are running as a regular non-root user and install to user directories instead - `arkade` should do the same (by default??) on systems where a `~/bin/` or `~/.local/bin` directory exists and that is also in `$PATH`.

Additionally, if a writeable directory in `$PATH` exists - `arkade` should be made to prefer installing tools it installs there over `~/.arkade/bin` in keeping with the principle of least astonishment - some users may already have pre-existing tools in say `~/.local/bin` which are likely to be preferred over those in `~/.arkade/bin` - especially if this directory is appended to `$PATH` as per instructions.

## Possible Solution

IMO, `arkade` has no real reason to be installed to the system directories if it perfectly well can be executed being installed to the a location in the regular non-root user's `$PATH`.  This is also in keeping with the principle of least privilege.  IMO (and this is my biased opinion off course) the default should change to only install to system directories if explicitly requested by the user (or when none of the user's directories exist in `$PATH`).  This ensures `arkade` runs in more situations than it currently does - with fewer post-install manual steps required to be done by the user.

## Steps to Reproduce (for bugs)

1. I did not install `arkade` to `/usr/local/bin/arkade`, instead I installed it to `~/.local/bin/arkade`
2. I had tools like `kubectl`, `hugo` and `terraform` already installed to `~/.local/bin`
3. Not thinking critically about things, I naively followed `arkade get terraform` instructions to `export PATH=$PATH:$HOME/.arkade/bin/`
4. I invoked `terraform` expecting the newer version to be used but spent 20 or so minutes troubleshooting issues in `terraform apply` only to find out the older version (i.e. the one from `~/.local/bin/terraform`) was preferred.
5. I ended up doing a `mv ~/.arkade/bin/terraform` to `~/.local/bin/` -- which seems to defeat the purpose of `arkade` providing a quick, almost-zero-configuration experience for installing tools.

## Context

This is my current "workaround" to ensure `arkade` and the tools it installs are placed in `~/.local/bin`

```
install-arkade() {
( cd "$TMPDIR";
  curl -sfLS https://dl.get-arkade.dev | sh
  install -m775 arkade ~/.local/bin/arkade
  rm arkade

  mkdir -p ~/.arkade/.helm/

  if ! test -L ~/.arkade/bin; then
    mv ~/.arkade/bin/* ~/.local/bin/
    rmdir ~/.arkade/bin/
    ln -svf ~/.local/bin/ ~/.arkade/bin       # This ensures arkade manages existing installations of tools
  fi
)
}
```

## Your Environment

* What Kubernetes distribution are you using?

```
kubectl version
```

* Operating System and version (e.g. Linux, Windows, MacOS):

```
uname -a
Linux leto 5.10.0-3-amd64 #1 SMP Debian 5.10.12-1 (2021-01-30) x86_64 GNU/Linux

echo $PATH
/home/user/.local/bin:/home/user/.bin:/home/user/.cargo/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games:/usr/local/sbin:/usr/sbin:/sbin:/usr/lib/pm-utils/bin
```

* What arkade version is this?

```
arkade version
...
Version: 0.7.10
Git Commit: 0931a3af1754c0e5fdd4013b666cdac2821a300d
```

**Comment by alexellis** [comment]
Hi @shalomb thanks for showing an interest in this project.

I'm a little confused when you say "don't install to system directories", because:

1) Arkade's installer when run as a normal user, won't install to a system directory
2) Arkade never writes to system directories.

I'd love to help you. Can you explain what I'm missing?

**Comment by shalomb** [comment]
@alexellis - Thanks for `arkade` and the effort put it to it - I can see it's use as a time-saving/convenience tool but it's not been a smooth user-experience for me so far.

I would have hoped I described the issue well enough already, it's not a question but a suggestion for improvement (to avoid an actual issue) on the user experience.  Let me try again.

## Issue 1 - Complete the installation after arkade is downloaded

```
Γ curl -sLS https://dl.get-arkade.dev | sh;
x86_64
Downloading package https://github.com/alexellis/arkade/releases/download/0.7.10/arkade as /tmp/arkade
Download complete.
============================================================
The script was run as a user who is unable to write
to /usr/local/bin. To complete the installation the
following commands may need to be run manually.
============================================================
sudo cp arkade /usr/local/bin/arkade
sudo ln -sf /usr/local/bin/arkade /usr/local/bin/ark
```

 `arkade` is suggesting here that the installation is *not* complete (and it isn't complete) and really is expecting to be installed to `/usr/local/bin/arkade`.  If the user has one of his directories in `$PATH` - why doesn't `arkade` prefer one of those and install itself there? The case for improvement on this front is convention/zero post-install configuration - that does not require the user to do anything more than the installation instructions provide. i.e. why do I need to `cp -av arkade ~/bin/` to put it in `$PATH` - it's implied that I've asked arkade to do this when I invoked `curl -sLS https://dl.get-arkade.dev | sh`.

As a a non-root user, I expect `curl -sLS https://dl.get-arkade.dev | sh;` to install `arkade` so that it can be invoked as a command later.

## Issue 2 - Reuse users ~/bin or ~/.local/bin directories as the destination for the tools that arkade installs

This applies to any tool/binary that arkade install but let's just take `terraform` as an example

```
Γ arkade get terraform
Downloading terraform
...
# Add (terraform) to your PATH variable
export PATH=$PATH:$HOME/.arkade/bin/
...
# Add (terraform) to your PATH variable
export PATH=$PATH:$HOME/.arkade/bin/
...
```

- If the user has an existing `terraform` installation (lets say managed by other non-arkade means and let's say installed to `~/bin/terraform`) - these instructions don't ensure that the new version of terraform is the one invoked and the *existing one still takes precedence* (by virtue of `$HOME/.arkade/bin/` being *appended* to `$PATH`). This is a problem because the user invoking `terraform` isn't necessarily aware that the existing installation is used -  and then goes on an epic wild-goose chase debugging things in the terraform code - whereas the issue was that the version of terraform invoked was incorrect (and this is because `arkade` doesn't manage the existing installation - when arguably it should).

- To fix this the user has to make choices that arkade isn't aware of 
  - either delete the existing `terraform` installation
  -  play with the `$PATH` env.var and place/prepend `$HOME/.arkade/bin/` prior to any other path where `terraform` may reside 
  - both of these cases defeat `arkade`'s _raison d'etre_ and are inconvenient extra post-install config that can/should be managed by `arkade`. How? `arkade` can detect existing installations (e.g. `type -a terraform`) and update them in-place or warn the user. I'd argue that by using one of `$HOME/.local/bin/` or `$HOME/bin/` instead of `$HOME/.arkade/bin/` these issues disappear entirely.

The case for improvement here is to ensure that `arkade get $tool` sets the system up so that the user gets the `$tool` (and importantly version) that was requested.

As a non-root user, I would like `arkade` to install tools in a manner that invoking them as commands afterwards *always ensures* the ones invoked are the ones that arkade installs.

## My workaround

```
install-arkade() {
( cd "$TMPDIR";
  curl -sfLS https://dl.get-arkade.dev | sh
  install -m775 arkade ~/.local/bin/arkade
  rm arkade

  mkdir -p ~/.arkade/.helm/

  if ! test -L ~/.arkade/bin; then
    mv ~/.arkade/bin/* ~/.local/bin/
    rmdir ~/.arkade/bin/
    ln -svf ~/.local/bin/ ~/.arkade/bin       # This ensures arkade manages existing installations of tools
  fi
)
}
```

- This function puts `arkade` into `~/.local/bin/arkade` (since `~/.local/bin/` is in my `$PATH`) - this ensures `arkade` is installed and is invokable as a command.
- Ensures `~/.arkade/bin` is symlinked to `~/.local/bin` - this ensures that if `arkade` installs a tool, it will *overwrite* any that may already exist in `~/.local/bin` so that there is *only one version of the `$tool`* - the desired outcome (and arguably arkade's goal).
